Associate restore snapshot task to parent mount task

nicktindall opened the pull request:
Implemented the test as an integration test that polls the tasks API to confirm the association of parent/child tasks. Thanks for the feedback @DaveCTurner.

Closes #105830

nicktindall committed:
Associate restore snapshot task to parent mount task

Closes #105830

nicktindall commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java:
```diff
@@ -237,33 +237,32 @@ protected void masterOperation(
+                // Log snapshot restore at the DEBUG log level
+                .quiet(true);
+            // Specify the mount task as the parent of the refresh task
+            restoreSnapshotRequest.setParentTask(clusterService.localNode().getId(), task.getId());
```
`setParentTask` doesn't follow the fluent pattern the other setters do, so this had to be separate

DaveCTurner replied:
You could use a `org.elasticsearch.client.internal.ParentTaskAssigningClient` if you really wanted a one-liner here.

nicktindall replied:
I think it's probably overkill in this case, but thanks!

nicktindall commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotActionTests.java:
```diff
@@ -0,0 +1,190 @@
+    private static final String SNAPSHOT_INDEX_NAME = "snapshotIndexName";
+    private static final IndexId INDEX_ID = new IndexId(SNAPSHOT_INDEX_NAME, UUID.randomUUID().toString());
+    private static final String SNAPSHOT_NAME = "snapshotName";
+    private static final SnapshotId SNAPSHOT_ID = new SnapshotId(SNAPSHOT_NAME, UUID.randomUUID().toString());
```
I had to use `UUID.randomUUID().toString()` here instead of `randomUUID()` because of a security context error. This is the first time I've used the LuceneTest methods so open to suggestions if this is wrong.

DaveCTurner replied:
Randomness isn't available in `static` context here, but there's no need for this constant to be `static`.

nicktindall replied:
Ahh this is because of the reproducible randomness thing (being in the context of the seed). Makes sense!

DaveCTurner reviewed (dismissed):
Yeah mocking this much stuff is pretty bad IMO, it will get in the way of future changes/refactorings to these APIs I'm sure.

I would suggest doing this as a `org.elasticsearch.xpack.searchablesnapshots.BaseSearchableSnapshotsIntegTestCase` which gives you real implementations of all this stuff. Submit a task to the master which blocks, something like this:

https://github.com/elastic/elasticsearch/blob/69ff7dfb2010872381f2f45ffc96c83595fb5672/server/src/test/java/org/elasticsearch/cluster/routing/BatchedRerouteServiceTests.java#L89-L101

(get the `clusterService` using `internalCluster().getCurrentMasterNodeInstance(ClusterService.class)`)

That will give you time to call the list-tasks API and verify that the mount task is the parent of the restore task.

nicktindall committed:
Implement as integration test instead of unit test

nicktindall commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -1155,6 +1163,75 @@ private void assertSearchableSnapshotStats(String indexName, boolean cacheEnable
+                    .filter(t -> t.action().equals(MountSearchableSnapshotAction.NAME))
+                    .findAny()
+                    .orElseThrow(() -> new AssertionError("No mount snapshot task found"));
+                assertEquals(mountSnapshotTask.taskId(), restoreSnapshotTask.parentTaskId());
```
There's no point retrying here if this fails, I'll move that out of the `assertBusy` somehow. Perhaps `assertBusy` isn't appropriate here, I just wanted to avoid polling the list tasks API forever.

nicktindall committed:
Tidy up polling logic

nicktindall commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -1155,6 +1164,78 @@ private void assertSearchableSnapshotStats(String indexName, boolean cacheEnable
+            waitUntil(() -> {
+                restoreSnapshotTask[0] = getTaskForActionFromMaster(TransportRestoreSnapshotAction.TYPE.name());
+                return restoreSnapshotTask[0] != null;
+            });
```
This still looks a bit janky, if there's a more suitable pattern about please advise. The idea was to just limit how long we poll the tasks API for in the event something goes wrong.

DaveCTurner replied:
I think just sticking an `assertBusy` around the whole thing would be fine. We only wait 10s if something goes wrong. It's true that any other failures in here will also cause `assertBusy` to retry but that's not really the point of this test, we have other tests to check these things.

ywangd replied:
I agree waiting a bit here does not concern me. But if we really try to avoid polling, it is possible to achieve so by having a `ClusterStateObserver` that triggers and blocks when the index is assigned. We can be sure the tasks are already registered when that happens, i.e. something like the follows

```java
final var clusterStateObserver = new ClusterStateObserver(clusterService, null, logger, new ThreadContext(Settings.EMPTY));
final var indexAssignedBarrier = new CyclicBarrier(2);
clusterStateObserver.waitForNextChange(new ClusterStateObserver.Listener() {
    @Override
    public void onNewClusterState(ClusterState state) {
        safeAwait(indexAssignedBarrier);
        safeAwait(indexAssignedBarrier);
    }
    ...
}, state -> state.routingTable().hasIndex(indexName));
```

DaveCTurner replied:
Ah yeah nice idea.

nicktindall replied:
I went ahead with just using `assertBusy` for readability. To do this with the observer, I think we'd need to add the observer, submit the mount task, then wait on the observer which would mean it'd all need to be inline so probably not worth it given we're not too fussed about polling. Perhaps if there's value in it we could add something like `T ESTest#doSomethingThenWaitForIndexCreation(Supplier<T> action)` down the track.

nicktindall committed:
Remove dead code

elasticsearchmachine commented:
Hi @nicktindall, I've created a changelog YAML for you.

nicktindall committed:
Update docs/changelog/108705.yaml

nicktindall committed:
Wait for mount to complete, delete index at end of test

nicktindall commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -1155,6 +1164,78 @@ private void assertSearchableSnapshotStats(String indexName, boolean cacheEnable
+            safeAwait(cyclicBarrier);   // Unblock the master thread
+            if (response != null) {
+                response.actionGet();
+                assertAcked(indicesAdmin().prepareDelete(indexName));
```
The `actionGet` and delete index here were an attempt to fix [a failure that occurred in CI in a teardown](https://gradle-enterprise.elastic.co/s/762xzz33sxoqg/tests/task/:x-pack:plugin:searchable-snapshots:internalClusterTest/details/org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsIntegTests/testMountingSnapshotLinksRefreshTaskAsChild?expanded-stacktrace=WyIwLTEiXQ&top-execution=1), one or both of these may be necessary. Happy to remove the delete if it's not necessary.

DaveCTurner replied:
Nice catch, yeah, it seems we have to wait for the mount to complete before finishing the test or else there's a race between cleaning up the test cluster and the mount process creating the internal `.snapshot-blob-cache` index.

I expect there's no need to manually delete the test index here tho but it doesn't hurt either. Would you add a code comment about why we need these nonobvious cleanup steps?

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -1155,6 +1164,78 @@ private void assertSearchableSnapshotStats(String indexName, boolean cacheEnable
+        if (response.getTasks().isEmpty()) {
+            return null;
+        }
+        return response.getTasks().get(0);
```
If we run this whole thing in an `assertBusy` then we can just assert that there's exactly one task in the response to trigger the needed retries.

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -1155,6 +1164,78 @@ private void assertSearchableSnapshotStats(String indexName, boolean cacheEnable
+        } finally {
+            safeAwait(cyclicBarrier);   // Unblock the master thread
+            if (response != null) {
+                response.actionGet();
```
This'd be as good a time as any to stop using these infinite waits in tests, doing something like this and then using `safeGet()` instead:

```diff
diff --git a/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java b/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
index 80f9f2abea1..94c89318718 100644
--- a/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
@@ -2154,6 +2154,10 @@ public abstract class ESTestCase extends LuceneTestCase {
     public static <T> T safeAwait(SubscribableListener<T> listener) {
         final var future = new PlainActionFuture<T>();
         listener.addListener(future);
+        return safeGet(future);
+    }
+
+    private static <T> T safeGet(Future<T> future) {
         try {
             return future.get(SAFE_AWAIT_TIMEOUT.millis(), TimeUnit.MILLISECONDS);
         } catch (InterruptedException e) {
```

DaveCTurner reviewed (commented):
Test looks great. I left a few pointers but nothing substantial.

nicktindall committed:
Tidy up test
- use assertBusy to reduce task-get complexity
- don't wait forever for mount to complete
- document additional required cleanup
- add ESTestCase#safeGet and use it

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

nicktindall commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -2154,15 +2155,19 @@ public static void safeAcquire(Semaphore semaphore) {
+            throw new AssertionError("safeGet: listener was completed exceptionally", e);
         } catch (TimeoutException e) {
-            throw new AssertionError("safeAwait: listener was not completed within the timeout", e);
+            throw new AssertionError("safeGet: listener was not completed within the timeout", e);
```
I didn't add a test for this. There doesn't seem to be any tests for any of the `safeXXX` methods in ESTestCaseTests`. Happy to add if we think there's value in it.

DaveCTurner replied:
No need IMO, it's simple enough for test code. I can't think of a way this could be wrong and still compile.

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -1155,6 +1163,74 @@ private void assertSearchableSnapshotStats(String indexName, boolean cacheEnable
+    }
+
+    private TaskInfo getTaskForActionFromMaster(String action) {
+        var ltr = new ListTasksRequest().setDetailed(true).setNodes(internalCluster().getMasterName()).setActions(action);
```
Would rather not use abbrs. for var. names wherever poss. :)

Here I think I'd just inline the variable, it's only used in one place - saves having to think up a name at all.

DaveCTurner commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java:
```diff
@@ -1155,6 +1163,74 @@ private void assertSearchableSnapshotStats(String indexName, boolean cacheEnable
+        var ltr = new ListTasksRequest().setDetailed(true).setNodes(internalCluster().getMasterName()).setActions(action);
+        ListTasksResponse response = client().execute(TransportListTasksAction.TYPE, ltr).actionGet();
+        int matchingTasks = response.getTasks().size();
+        assertEquals(String.format(Locale.ROOT, "Expected a single task for action %s, got %d", action, matchingTasks), 1, matchingTasks);
```
I'd use `assertThat(..., hasSize(1))` here - that should give the same information on failure as what you've got here but without needing a hand-crafted message.

DaveCTurner reviewed (commented):
Looks good, just a few tiny style points.

nicktindall committed:
Tidy up from feedback

DaveCTurner reviewed (approved):
LGTM
